core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -0,0 +1,669 @@
+
+  init()
+
+  /**
+   * Perform any startup operations.
+   *
+   * This includes declaring this instance as the cache to use in the
+   * [[ApplicationCacheCheckFilterRelay]].
+   */
+  private def init(): Unit = {
+    ApplicationCacheCheckFilterRelay.setApplicationCache(this)
+  }
+
+  /**
+   * Stop the cache.
+   * This will reset the relay in [[ApplicationCacheCheckFilterRelay]].
+   */
+  def stop(): Unit = {
+    ApplicationCacheCheckFilterRelay.resetApplicationCache()
+  }
+
+  /**
+   * Get an entry.
+   *
+   * Cache fetch/refresh will have taken place by the time this method returns.
+   * @param appAndAttempt application to look up in the format needed by the history server web UI,
+   *                      `appId/attemptId` or `appId`.
+   * @return the entry
+   */
+  def get(appAndAttempt: String): SparkUI = {
+    val parts = splitAppAndAttemptKey(appAndAttempt)
+    get(parts._1, parts._2)
+  }
+
+  /**
+   * Get the Spark UI, converting a lookup failure from an exception to `None`.
+   * @param appAndAttempt application to look up in the format needed by the history server web UI,
+   *                      `appId/attemptId` or `appId`.
+   * @return the entry
+   */
+  def getSparkUI(appAndAttempt: String): Option[SparkUI] = {
+    try {
+      val ui = get(appAndAttempt)
+      Some(ui)
+    } catch {
+      case NonFatal(e) => e.getCause() match {
+        case nsee: NoSuchElementException =>
+          None
+        case cause: Exception => throw cause
+      }
+    }
+  }
+
+  /**
+   * Get the associated spark UI.
+   *
+   * Cache fetch/refresh will have taken place by the time this method returns.
+   * @param appId application ID
+   * @param attemptId optional attempt ID
+   * @return the entry
+   */
+  def get(appId: String, attemptId: Option[String]): SparkUI = {
+    lookupAndUpdate(appId, attemptId)._1.ui
+  }
+
+  /**
+   * Look up the entry; update it if needed.
+   * @param appId application ID
+   * @param attemptId optional attempt ID
+   * @return the underlying cache entry -which can have its timestamp changed, and a flag to
+   *         indicate that the entry has changed
+   */
+  private def lookupAndUpdate(appId: String, attemptId: Option[String]): (CacheEntry, Boolean) = {
+    metrics.lookupCount.inc()
+    val cacheKey = CacheKey(appId, attemptId)
+    var entry = appCache.getIfPresent(cacheKey)
+    var updated = false
+    if (entry == null) {
+      // no entry, so fetch without any post-fetch probes for out-of-dateness
+      // this will trigger a callback to loadApplicationEntry()
+      entry = appCache.get(cacheKey)
+    } else if (!entry.completed) {
+      val now = clock.getTimeMillis()
+      if (now - entry.probeTime > refreshInterval) {
+        log.debug(s"Probing at time $now for updated application $cacheKey -> $entry")
+        metrics.updateProbeCount.inc()
+        updated = time(metrics.updateProbeTimer) {
+          entry.updateProbe()
```

Comment:
right, I guess I just wanted to point out that with the changes here, this probe is entirely independent from replay.  Replay happens with normal log-checking -- that frequency is controlled by `spark.history.fs.update.interval`.   Here, we're just checking whether that regular log scanning has already loaded an updated UI for this attempt, and that is it.  Since `spark.history.fs.update.interval` is entirely controlling the expensive part, we may not need any other interval.